src/gpio/mod.rs

```diff
@@ -331,5 +333,45 @@ macro_rules! define_led {
     }
 }
 
+/// Opaque AFRL register
+pub struct AFRL<GPIO>(PhantomData<GPIO>);
+/// Opaque AFRH register
+pub struct AFRH<GPIO>(PhantomData<GPIO>);
+/// Opaque MODER register
+pub struct MODER<GPIO>(PhantomData<GPIO>);
+/// Opaque OTYPER register
+pub struct OTYPER<GPIO>(PhantomData<GPIO>);
+/// Opaque PUPDR register
+pub struct PUPDR<GPIO>(PhantomData<GPIO>);
+
+impl_parts!(
+    GPIOA, gpioa;
+    GPIOB, gpiob;
+    GPIOC, gpioc;
+    );
+
+//Each I/O pin (except PH3 for STM32L496xx/4A6xx devices) has a multiplexer with up to
+//sixteen alternate function inputs (AF0 to AF15) that can be configured through the
+//GPIOx_AFRL (for pin 0 to 7) and GPIOx_AFRH (for pin 8 to 15) registers
+impl_gpio!(A, GPIOA, gpioaen, gpioarst,
+           AFRL: [PA0, 0; PA1, 1; PA2, 2; PA3, 3; PA4, 4; PA5, 5; PA6, 6; PA7, 7;],
+           AFRH: [PA8, 8; PA9, 9; PA10, 10; PA11, 11; PA12, 12; PA13, 13; PA14, 14; PA15, 15; ]
+          );
+impl_gpio!(B, GPIOB, gpioben, gpiobrst,
+           AFRL: [PB0, 0; PB1, 1; PB2, 2; PB3, 3; PB4, 4; PB5, 5; PB6, 6; PB7, 7;],
+           AFRH: [PB8, 8; PB9, 9; PB10, 10; PB11, 11; PB12, 12; PB13, 13; PB14, 14; PB15, 15; ]
+          );
+impl_gpio!(C, GPIOC, gpiocen, gpiocrst,
+           AFRL: [PC0, 0; PC1, 1; PC2, 2; PC3, 3; PC4, 4; PC5, 5; PC6, 6; PC7, 7;],
+           AFRH: [PC8, 8; PC9, 9; PC10, 10; PC11, 11; PC12, 12; PC13, 13; PC14, 14; PC15, 15; ]
+          );
+
 #[cfg(feature = "STM32L476VG")]
 pub mod stm32l476vg;
+#[cfg(feature = "STM32L476VG")]
+pub use gpio::stm32l476vg::gpio::*;
```

Comment:
I suggest to not re-export board specific pins to ensure that there will be clear distinction between generic and specific ones

This is also important because I'd like to generate docs for every feature so that user would see it on docs.rs
With re-export compilation shall fail

Reply:
I'm not sure how to solve this. My goal is actually to make common pins and device-specific pins indistinguishable, so I can do:
```rust
use hal::gpio;

let gpiob = gpio::A::new(...); // "common"
let gpiof = gpio::F::new(...); // "specific"
```

Otherwise, I have to keep track of what's common versus not. This makes swapping parts harder, since other chips may not have such a "common" vs "specific" distinction, and would only have a single `gpio` namespace. Eventually, I'd like to have a def for every chip, and make "common" ports private. Would doing that immediately be better?

Reply:
Hm I think it might be good to go with pattern of re-exporting common pins into specific module.
This way you can have it as:
`use hal::gpio::<board_module> as gpio`

I'd say let's leave common pins public in `hal::gpio` too but for convenience sake they will be re-exported to board specific module.
This way there is no way to introduce conflict for docs.rs and your goal of having single gpio module is achieved too.

This is what you have in mind?

Reply:
Yeah, that would work well (So instead, I can do
```rust
use hal::gpio::stm32l496ag::gpio as gpio
```

The only problem is that I can't figure out how to do that and not break doc compilation. The problem is the `pub(crate)` defns in the root that are created by `impl_parts!`. I'll push another (broken) update so you can see.